Workflow/Tests/WorkflowObserverTests.swift

```diff
@@ -0,0 +1,537 @@
+        observer2.onDidReceiveAction = { _, _, _ in
+            callbackSequence.append("two")
+        }
+
+        let chained = ChainedWorkflowObserver(observers: [observer1, observer2])
+        chained.workflowDidReceiveAction(
+            StateTransitioningWorkflow.Event.toggle,
+            workflow: StateTransitioningWorkflow(),
+            session: .testingSession
+        )
+
+        XCTAssertEqual(
+            callbackSequence,
+            ["one", "two"]
+        )
+    }
+
+    func test_chainedObserver_willApplyActionCallbacks() {
+        var willApplyCallbackSequence: [String] = []
+        var didApplyCallbackSequence: [String] = []
+
+        let observer1 = TestObserver()
+        observer1.onApplyAction = { _, _, _, _ in
+            willApplyCallbackSequence.append("one")
+            return { _, _ in
+                didApplyCallbackSequence.append("one")
+            }
+        }
+
+        let observer2 = TestObserver()
+        observer2.onApplyAction = { _, _, _, _ in
+            willApplyCallbackSequence.append("two")
+            return { _, _ in
+                didApplyCallbackSequence.append("two")
+            }
+        }
+
+        let chained = ChainedWorkflowObserver(observers: [observer1, observer2])
+        let didApply = chained.workflowWillApplyAction(
+            StateTransitioningWorkflow.Event.toggle,
+            workflow: StateTransitioningWorkflow(),
+            state: false,
+            session: .testingSession
+        )
+
+        XCTAssertEqual(willApplyCallbackSequence, ["one", "two"])
+        XCTAssertEqual(didApplyCallbackSequence, [])
+
+        didApply?(true, nil)
+
+        XCTAssertEqual(didApplyCallbackSequence, ["two", "one"])
+    }
+}
+
+// MARK: - Integration Points
+
+extension WorkflowObserverTests {
+    func test_workflowHost_chainsObservers() {
+        // 0 observers
+        do {
+            let host = WorkflowHost(
+                workflow: Parent(),
+                observers: nil
+            )
+
+            XCTAssertNil(host.rootNode.observer)
+        }
+
+        // 1 observer
+        do {
+            let host = WorkflowHost(
+                workflow: Parent(),
+                observers: [TestObserver()]
+            )
+
+            XCTAssert(host.rootNode.observer is TestObserver)
+        }
+
+        // > 1 observer
+        do {
+            struct NoOpObserver: WorkflowObserver {}
+
+            let host = WorkflowHost(
+                workflow: Parent(),
+                observers: [TestObserver(), NoOpObserver()]
+            )
+
+            let observer = host.rootNode.observer as? ChainedWorkflowObserver
+            XCTAssertNotNil(observer)
+
+            XCTAssertEqual(observer?.observers.count, 2)
+            XCTAssert(observer?.observers.first is TestObserver)
+            XCTAssert(observer?.observers.last is NoOpObserver)
+        }
+    }
+}
+
+// MARK: - Utilities
+
+private final class TestObserver: WorkflowObserver {
```

Comment:
that's true. i think i had something like that at one point, but decided against it since some of the type checking assertions then ended up being done by string comparisons which felt kinda awkward/unwieldy (or at least that's the approach i came up with at the time).